### ISSUE TITLE
How to use tensor in Ax service API

### ISSUE THREAD

**Issue by Wang-Zujian** [facebook/Ax#2496]
Hi，everyone！
     When I set up my optimization with the service API, the CPU is always working, and my GPU is not working, is it because the functions I define that need to be optimized are all variables defined by numpy？ And how should I call the GPU to speed up my optimization?
      
> When I use tensor to define my variables, I encountered the problem below。

![1717413868194](https://github.com/facebook/Ax/assets/147903336/5f9aea0c-c662-4084-851e-12fd0326d550)

> I have solved the above error, but I checked my GPU operation in the background manager 0

      

> Here are some of my code.
```
    def PB_CostFunc(parameterization):
          for i in range(10):
              parameterization.get(f"x{i + 1}")

          d1 = parameterization['d1']
          d2 = parameterization['d2']
          d3 = parameterization['d3']
          d4 = parameterization['d4']
          d5 = parameterization['d5']
          phi1 = parameterization['phi1']
          phi2 = parameterization['phi2']
          phi3 = parameterization['phi3']
          phi4 = parameterization['phi4']
          phi5 = parameterization['phi5']
      
          device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
          torch.set_default_dtype(torch.double)
      
          dz0 = 10
      
          JN_11, JN_12, JN_21, JN_22 = lcs.Jones_2by2_FastGPU(k0, ne, no, DZ, PHI, 0, phi0, theta_in)
      
          J01 = torch.ones(N, device=device)
          J02 = (1j * torch.ones(N, device=device)) * torch.cos(theta_in)
      
          JN1 = JN_11 * J01 + JN_12 * J02
          JN2 = JN_21 * J01 + JN_22 * J02
          JN2 = JN2 / torch.cos(theta_in)
          JN_norm = torch.sqrt(torch.abs(JN1) ** 2 + torch.abs(JN2) ** 2)
          JN1 = JN1 / JN_norm
          JN2 = JN2 / JN_norm
      
          NN1 = N_ii * N_jj
          NN2 = N_kk
      
          JN1 = JN1.reshape((NN2, NN1))
          JN2 = JN2.reshape((NN2, NN1))
          F = phi0.reshape((NN2, NN1))
          F = torch.exp(1j * F * 2)
          A1 = torch.sum(F * JN1, dim=-2) / NN2
          A2 = torch.sum(F * JN2, dim=-2) / NN2
          Eff = torch.abs(A1) ** 2 + torch.abs(A2) ** 2
          cost = 1 - torch.mean(Eff)
          cost = cost.cpu().numpy().reshape(1,-1)[0][0]
      
          return {"PB_CostFunc": cost}
      
    ax_client = AxClient(enforce_sequential_optimization = False)
    ax_client.create_experiment(
        name="PB_optimization",
        parameters=[
            {
                "name": "d1",
                "type": "range",
                "bounds": [100.0, 3000.0],
                "value_type": "float",  # Optional, defaults to inference from type of "bounds".
                "log_scale": False,  # Optional, defaults to False.
            },
            {
                "name": "d2",
                "type": "range",
                "bounds": [100.0, 3000.0],
                "value_type": "float",
            },
            {
                "name": "d3",
                "type": "range",
                "bounds": [100.0, 3000.0],
                "value_type": "float",
            },
            {
                "name": "d4",
                "type": "range",
                "bounds": [100.0, 3000.0],
                "value_type": "float",
            },
            {
                "name": "d5",
                "type": "range",
                "bounds": [100.0, 3000.0],
                "value_type": "float",
            },
            {
                "name": "phi1",
                "type": "range",
                "bounds": [-np.pi, np.pi],
                "value_type": "float",
            },
            {
                "name": "phi2",
                "type": "range",
                "bounds": [-np.pi, np.pi],
                "value_type": "float",
            },
            {
                "name": "phi3",
                "type": "range",
                "bounds": [-np.pi, np.pi],
                "value_type": "float",
            },
            {
                "name": "phi4",
                "type": "range",
                "bounds": [-np.pi, np.pi],
                "value_type": "float",
            },
            {
                "name": "phi5",
                "type": "range",
                "bounds": [-np.pi, np.pi],
                "value_type": "float",
            },
    
        ],
        objectives={"PB_CostFunc": ObjectiveProperties(minimize=True)},
        choose_generation_strategy_kwargs={"max_parallelism_override": 5},
    )
    
    sol = np.zeros([10, 10])
    val = np.zeros([10, 1])
    j = 0
    
    for i in range(10000):
        parameters, trial_index = ax_client.get_next_trial()
        # Local evaluation here can be replaced with deployment to external system.
        ax_client.complete_trial(
            trial_index=trial_index, raw_data=PB_CostFunc(parameters)
        )
```

**Comment by saitcakmak** [comment]
Hi @Wang-Zujian. For `AxClient` to utilize the GPU, you need to specify the `torch_device` argument when constructing `AxClient`. In the code sample provided, only the `PB_CostFunc` will utilize the GPU. You can replace
```
ax_client = AxClient(enforce_sequential_optimization = False)
```
with
```
ax_client = AxClient(enforce_sequential_optimization = False, torch_device=torch.device("cuda" if torch.cuda.is_available() else "cpu"))
```

**Comment by Wang-Zujian** [comment]
> Hi @Wang-Zujian. For `AxClient` to utilize the GPU, you need to specify the `torch_device` argument when constructing `AxClient`. In the code sample provided, only the `PB_CostFunc` will utilize the GPU. You can replace
> 
> ```
> ax_client = AxClient(enforce_sequential_optimization = False)
> ```
> 
> with
> 
> ```
> ax_client = AxClient(enforce_sequential_optimization = False, torch_device=torch.device("cuda" if torch.cuda.is_available() else "cpu"))
> ```

Thank you very much for your prompt reply. Maybe my calculation amount is relatively small. Now the GPU utilization rate can reach about 6%. If I want to make better use of my GPU performance, can I modify the numbers with the following code to increase my parallelism?
` choose_generation_strategy_kwargs={"max_parallelism_override": 5`

**Comment by saitcakmak** [comment]
It is expected to see small GPU utilization, since BO utilizes relatively small tensors (unless you run many iterations). GPU will have limited impact on runtime with Ax in most typical use cases. 

The `max_parallelism_override` does not have any effect on the underlying computations. It controls how many trials can be run in parallel, and is typically leveraged when using `Scheduler`.
